clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -471,40 +471,45 @@ inlineNonRepWorker e@(Case scrut altsTy alts)
     isInlined <- zoomExtra (alreadyInlined f cf)
     limit     <- Lens.use (extra.inlineLimit)
     tcm       <- Lens.view tcCache
-    let scrutTy = termType tcm scrut
-        noException = not (exception tcm scrutTy)
-    if noException && (Maybe.fromMaybe 0 isInlined) > limit
-      then
-        trace (concat [ $(curLoc) ++ "InlineNonRep: " ++ showPpr (varName f)
-                      ," already inlined " ++ show limit ++ " times in:"
-                      , showPpr (varName cf)
-                      , "\nType of the subject is: " ++ showPpr scrutTy
-                      , "\nFunction " ++ showPpr (varName cf)
-                      , " will not reach a normal form, and compilation"
-                      , " might fail."
-                      , "\nRun with '-fclash-inline-limit=N' to increase"
-                      , " the inlining limit to N."
-                      ])
-              (pure e)
-      else do
-        bodyMaybe   <- lookupVarEnv f <$> Lens.use bindings
-        nonRepScrut <- not <$> (representableType <$> Lens.view typeTranslator
-                                                  <*> Lens.view customReprs
-                                                  <*> pure False
-                                                  <*> Lens.view tcCache
-                                                  <*> pure scrutTy)
-        case (nonRepScrut, bodyMaybe) of
-          (True,Just b) -> do
-            Monad.when noException (zoomExtra (addNewInline f cf))
-
-            let scrutBody0 = mkTicks (bindingTerm b) (mkInlineTick f : ticks)
-            let scrutBody1 = mkApps scrutBody0 args
-
-            changed (Case scrutBody1 altsTy alts)
-
-          _ -> pure e
-  where
-    exception = isClassTy
+    let
+      scrutTy = termType tcm scrut
+
+      -- Constraint dictionary inlining always terminates, so we ignore the
+      -- usual inline safeguards.
+      notClassTy = not (isClassTy tcm scrutTy)
+      overLimit = notClassTy && (Maybe.fromMaybe 0 isInlined) > limit
+
+
+    bodyMaybe   <- lookupVarEnv f <$> Lens.use bindings
+    nonRepScrut <- not <$> (representableType <$> Lens.view typeTranslator
+                                              <*> Lens.view customReprs
+                                              <*> pure False
+                                              <*> Lens.view tcCache
+                                              <*> pure scrutTy)
+    case (nonRepScrut, bodyMaybe) of
+      (True, Just b) -> do
+        if overLimit then
+          trace ($(curLoc) ++ [I.i|
+            InlineNonRep: #{showPpr (varName f)} already inlined
+            #{limit} times in: #{showPpr (varName cf)}. The type of the subject
+            is:
+
+              #{showPpr scrutTy}
```

Comment:
@alex-mckenna Added a comment on this commit saying we should probably use `showPpr'` to prevent it not printing anything when using `CLASH_PPR_*`.

Reply:
I think a relatively harmless internal change might also be to make the `CLASH_PPR_*` flags only apply in certain cases, i.e. making `CLASH_PPR_TYPES` only come into effect when printing the type of a binder, or a type application. I think every other instance of a type _should_ be printed regardless of this flag - i.e. if I ask to print a type I should _never_ see an empty string

Reply:
That's a good idea, but for now I just used `showPpr'`